plugins/linkproviders/youtube.go

```diff
@@ -0,0 +1,158 @@
+			Description          string `json:"description"`
+			LiveBroadcastContent string `json:"liveBroadcastContent"`
+			Localized            struct {
+				Description string `json:"description"`
+				Title       string `json:"title"`
+			} `json:"localized"`
+			PublishedAt string `json:"publishedAt"`
+			Thumbnails  struct {
+				Default struct {
+					Height int    `json:"height"`
+					URL    string `json:"url"`
+					Width  int    `json:"width"`
+				} `json:"default"`
+				High struct {
+					Height int    `json:"height"`
+					URL    string `json:"url"`
+					Width  int    `json:"width"`
+				} `json:"high"`
+				Medium struct {
+					Height int    `json:"height"`
+					URL    string `json:"url"`
+					Width  int    `json:"width"`
+				} `json:"medium"`
+			} `json:"thumbnails"`
+			Title string `json:"title"`
+		} `json:"snippet"`
+	} `json:"items"`
+}
+
+func NewYoutubeProvider(b *bot.Bot, p *plugins.URLPlugin) error {
+	// Listen for youtube.com and youtu.be URLs
+	p.RegisterProvider("youtube.com", HandleYoutube)
+	p.RegisterProvider("youtu.be", HandleYoutube)
+
+	return nil
+}
+
+func HandleYoutube(b *bot.Bot, m *irc.Message, req *url.URL) bool {
+	// Get API key from seabird config
+	tc := &YoutubeConfig{}
+	err := b.Config("youtube", tc)
+	if err != nil {
+		return false
+	}
+
+	// Get the Video ID from the URL
+	p, _ := url.ParseQuery(req.RawQuery)
+	var id string
+	if len(p["v"]) > 0 {
+		// using full www.youtube.com/?v=bbq
+		id = p["v"][0]
+	}else {
+		// using short youtu.be/bbq
+		path := strings.Split(req.Path, "/")
+		if len(path) < 1{
+			return false
+		}
+		id = path[1]
+	}
+
+	// Get video duration and title
+	time, title := getVideo(id, tc.Key)
+
+	// Invalid video ID or no results
+	if time=="" && title=="" {
+		return false
+	}
+
+	// Send out the IRC message
+	msg := fmt.Sprintf("%s ~ %s", time, title)
+	b.Reply(m, "%s %s", youtubePrefix, msg)
+
+	return true
+}
+
+func getVideo(id string, key string) (time string, title string) {
+	// Build the API call
+	api := fmt.Sprintf("https://www.googleapis.com/youtube/v3/videos?part=contentDetails%%2Csnippet&id=%s&fields=items(contentDetails%%2Csnippet)&key=%s", id, key)
+
+	// Get the YouTube JSON
+	r, _ := http.Get(api)
+	defer r.Body.Close()
+
+	// Decode JSON into the Videos struct
+	var videos Videos
+	dec := json.NewDecoder(r.Body)
+	err := dec.Decode(&videos)
+	if err != nil {
+		return "", ""
+	}
+
+	// Make sure we found a video
+	if len(videos.Items) < 1 {
+		return "", ""
+	}
+
+	v := videos.Items[0]
+
+	// Format the duration
+	d, err := duration.FromString(v.ContentDetails.Duration)
+	if err != nil {
+		return "", ""
+	}
+
+	// always show minutes:seconds even if 00:01 or 1:00
+        dr := fmt.Sprintf("%02d:%02d", d.Minutes, d.Seconds)
+
+        // add Hours if not zero
+        if(d.Hours > 0) {
+                dr = fmt.Sprintf("%02d:%s", d.Hours, dr)
+        }
+        // add Days if not zero. overwrite all other time string
+        if(d.Days > 0){
```

Comment:
I could flip the logic so that I do something like

```go
if(d.Days > 0) {
        dr = // D:H:M:S
}else if(d.Hours > 0){
        dr = // H:M:S
}else{
        dr = // M:S
}
```